File: app/src/main/java/com/nollpointer/dates/views/SortCards.java
package com.nollpointer.dates.views;

import android.graphics.Color;
import android.support.v7.widget.CardView;
import android.view.View;

import com.nollpointer.dates.R;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SortCards implements SortCardsControl {

    private static final int MAX_COUNT = 3;

    private ArrayList<SortCardView> cards;
    private int[] currentSequence = new int[MAX_COUNT];
    private int[] answerSequence = new int[MAX_COUNT];

    private boolean isCheckMode = false;

    private int GREEN = Color.GREEN;
    private int RED = Color.RED;

    private View.OnClickListener listener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            int pos = getPosition(view.getId());
            onCardClick(pos);
        }
    };

    @Override
    public void setColors(int green, int red) {
        GREEN = green;
        RED = red;
    }

    public static SortCardsControl newInstance(View mainView) {
        SortCards sorts = new SortCards();
        sorts.cards = new ArrayList<>(MAX_COUNT);

        CardView cardView = mainView.findViewById(R.id.cardView0);
        sorts.cards.add(new SortCardView(cardView));
        cardView = mainView.findViewById(R.id.cardView1);
        sorts.cards.add(new SortCardView(cardView));
        cardView = mainView.findViewById(R.id.cardView2);
        sorts.cards.add(new SortCardView(cardView));

        sorts.initialize();
        return sorts;
    }

    private void initialize() {
        for (int i = 0; i < MAX_COUNT; i++) {
            SortCardView card = cards.get(i);
            card.setOnClickListener(listener);
            card.setNumber(i + 1);
            currentSequence[i] = i + 1;
        }
    }

    private void onCardClick(int position) {
        int sequenceNumber = currentSequence[position];
        sequenceNumber %= MAX_COUNT;
        sequenceNumber++;
        currentSequence[position] = sequenceNumber;
        cards.get(position).setNumber(sequenceNumber);
        if (isCheckMode)
            singleCardCheck(position);
    }


    @Override
    public void setQuestions(List<String> list) {
        for (int i = 0; i < MAX_COUNT; i++) {
            SortCardView card = cards.get(i);
            card.setBackgroundColor(Color.WHITE);
            card.setEvent(list.get(i));
            card.setNumber(i + 1);
            currentSequence[i] = i + 1;
        }
    }

    @Override
    public void setAnswerSequence(int[] sequence) {
        System.arraycopy(sequence, 0, answerSequence, 0, MAX_COUNT);
    }

    private int getPosition(int id) {
        int position = -1;
        switch (id) {
            case R.id.cardView0:
                position = 0;
                break;
            case R.id.cardView1:
                position = 1;
                break;
            case R.id.cardView2:
                position = 2;
                break;
        }
        return position;
    }

    @Override
    public void setCheckMode(boolean state) {
        isCheckMode = state;
    }

    @Override
    public boolean check() {
        boolean isCorrect = Arrays.equals(answerSequence, currentSequence);
        if (isCorrect) {
            for (SortCardView card : cards)
                card.setBackgroundColor(Color.GREEN);
        } else {
            for (int i = 0; i < MAX_COUNT; i++) {
                if (currentSequence[i] == answerSequence[i])
                    cards.get(i).setBackgroundColor(GREEN);
                else
                    cards.get(i).setBackgroundColor(RED);
            }
        }
        return isCorrect;
    }

    private void singleCardCheck(int position) {
        if (answerSequence[position] == currentSequence[position])
            cards.get(position).setBackgroundColor(GREEN);
        else
            cards.get(position).setBackgroundColor(RED);

    }
}

File: app/src/main/java/com/nollpointer/dates/sort/SortCards.kt
package com.nollpointer.dates.sort

import android.graphics.Color
import android.view.View
import androidx.cardview.widget.CardView
import java.util.*

class SortCards : SortCardsControl {
    private lateinit var cards: ArrayList<SortCardView>
    private val currentSequence = IntArray(MAX_COUNT)
    private val answerSequence = IntArray(MAX_COUNT)
    private var isCheckMode = false
    private var green = Color.GREEN
    private var red = Color.RED
    private val listener = View.OnClickListener { view ->
        val pos = getPosition(view.id)
        onCardClick(pos)
    }

    override fun setColors(g: Int, r: Int) {
        green = g
        red = r
    }

    private fun initialize() {
        for (i in 0 until MAX_COUNT) {
            val card = cards[i]
            card.setOnClickListener(listener)
            card.setNumber(i + 1)
            currentSequence[i] = i + 1
        }
    }

    private fun onCardClick(position: Int) {
        var sequenceNumber = currentSequence[position]
        sequenceNumber %= MAX_COUNT
        sequenceNumber++
        currentSequence[position] = sequenceNumber
        cards[position].setNumber(sequenceNumber)
        if (isCheckMode) singleCardCheck(position)
    }

    override fun setQuestions(list: List<String?>?) {
        for (i in 0 until MAX_COUNT) {
            val card = cards[i]
            card.setBackgroundColor(Color.WHITE)
            card.setEvent(list!![i])
            card.setNumber(i + 1)
            currentSequence[i] = i + 1
        }
    }

    override fun setAnswerSequence(sequence: IntArray?) {
        System.arraycopy(sequence!!, 0, answerSequence, 0, MAX_COUNT)
    }

    private fun getPosition(id: Int): Int {
        var position = -1
        when (id) {
            0 -> position = 0
            1 -> position = 1
            2 -> position = 2
        }
        return position
    }

    override fun setCheckMode(state: Boolean) {
        isCheckMode = state
    }

    override fun check(): Boolean {
        val isCorrect = answerSequence.contentEquals(currentSequence)
        if (isCorrect) {
            for (card in cards) card.setBackgroundColor(Color.GREEN)
        } else {
            for (i in 0 until MAX_COUNT) {
                if (currentSequence[i] == answerSequence[i]) cards[i].setBackgroundColor(green) else cards[i].setBackgroundColor(red)
            }
        }
        return isCorrect
    }

    private fun singleCardCheck(position: Int) {
        if (answerSequence[position] == currentSequence[position]) cards[position].setBackgroundColor(green) else cards[position].setBackgroundColor(red)
    }

    companion object {
        private const val MAX_COUNT = 3
        fun newInstance(mainView: View): SortCardsControl {
            val sorts = SortCards()
            sorts.cards = ArrayList(MAX_COUNT)
            var cardView: CardView = mainView.findViewById(0)
            sorts.cards.add(SortCardView(cardView))
            cardView = mainView.findViewById(0)
            sorts.cards.add(SortCardView(cardView))
            cardView = mainView.findViewById(0)
            sorts.cards.add(SortCardView(cardView))
            sorts.initialize()
            return sorts
        }
    }
}